File: pulsar/java/src/main/java/com/examples/hello/pulsar/consumers/ConsumerRunner.java
package com.examples.hello.pulsar.consumers;

import com.examples.hello.pulsar.services.PulsarRunner;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.pulsar.client.api.Consumer;

@Slf4j
@RequiredArgsConstructor
public class ConsumerRunner<T> extends PulsarRunner implements ConsumerRunnable {
  private final Consumer<T> consumer;
  private final MessageConsumer<T> sink;

  @Override
  @SneakyThrows
  protected void callOnce() {
    val msg = consumer.receive();
    sink.accept(msg);
    consumer.acknowledge(msg);
    log.debug("Received msg(time={} id={}), value={})", msg.getEventTime(), msg.getMessageId(), msg.getValue());
  }

  @Override
  public void close() throws Exception {
    stop();
    consumer.close();
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/consumers/ConsumerRunner.kt
package com.example.hello.pulsar.consumers

import com.example.hello.pulsar.services.PulsarRunner
import org.apache.pulsar.client.api.Consumer
import org.apache.pulsar.client.api.Message
import org.slf4j.LoggerFactory

class ConsumerRunner<T>(
  private val consumer: Consumer<T>,
  private val sink: (Message<T>) -> Unit
  ) : PulsarRunner(), ConsumerRunnable {

    override fun callOnce() {
        val msg = consumer.receive()
        sink.invoke(msg)
        consumer.acknowledge(msg)
        log.debug("Received msg(time={} id={}), value={})", msg.eventTime, msg.messageId, msg.value)
    }

    override fun close() {
        stop()
        consumer.close()
    }

  companion object {
    private val log = LoggerFactory.getLogger(ConsumerRunner::class.java)
  }
}
